compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/DescriptorMetadataSerializer.kt

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.backend.jvm.codegen
+
+import org.jetbrains.kotlin.backend.jvm.JvmBackendContext
+import org.jetbrains.kotlin.codegen.*
+import org.jetbrains.kotlin.codegen.binding.CodegenBinding
+import org.jetbrains.kotlin.codegen.serialization.JvmSerializationBindings
+import org.jetbrains.kotlin.codegen.serialization.JvmSerializerExtension
+import org.jetbrains.kotlin.ir.declarations.*
+import org.jetbrains.kotlin.ir.util.*
+import org.jetbrains.kotlin.metadata.jvm.serialization.JvmStringTable
+import org.jetbrains.kotlin.protobuf.MessageLite
+import org.jetbrains.kotlin.serialization.DescriptorSerializer
+import org.jetbrains.org.objectweb.asm.Type
+import org.jetbrains.org.objectweb.asm.commons.Method
+
+class DescriptorMetadataSerializer(
+    private val context: JvmBackendContext,
+    private val irClass: IrClass,
+    private val type: Type,
+    private val serializationBindings: JvmSerializationBindings,
+    parent: MetadataSerializer?
+) : MetadataSerializer {
+    private val serializerExtension = JvmSerializerExtension(serializationBindings, context.state, context.typeMapper)
+    private val serializer: DescriptorSerializer? =
+        when (val metadata = irClass.metadata) {
+            is DescriptorMetadataSource.Class -> DescriptorSerializer.create(
+                metadata.descriptor, serializerExtension, (parent as? DescriptorMetadataSerializer)?.serializer
+            )
+            is DescriptorMetadataSource.File -> DescriptorSerializer.createTopLevel(serializerExtension)
+            is DescriptorMetadataSource.Function -> DescriptorSerializer.createForLambda(serializerExtension)
+            else -> null
+        }
+
+    override val stringTable: JvmStringTable
+        get() = serializer!!.stringTable as JvmStringTable
+
+    override fun serialize(metadata: MetadataSource): MessageLite? {
+        val localDelegatedProperties = (irClass.attributeOwnerId as? IrClass)?.let(context.localDelegatedProperties::get)
+        if (localDelegatedProperties != null && localDelegatedProperties.isNotEmpty()) {
+            context.state.bindingTrace.record(
+                CodegenBinding.DELEGATED_PROPERTIES_WITH_METADATA,
+                // When serializing metadata for interfaces, `JvmSerializerExtension.serializeClass`
+                // looks at `$DefaultImpls` for some reason that's probably related to the old backend.
+                if (irClass.isInterface) context.typeMapper.mapClass(context.cachedDeclarations.getDefaultImplsClass(irClass)) else type,
+                localDelegatedProperties.mapNotNull { (it.owner.metadata as? DescriptorMetadataSource.LocalDelegatedProperty)?.descriptor }
```

Comment:
Given that this is descriptor-specific metadata serialization, I actually don't see why someone changed it from simply `it.descriptor` to this. (There was similar code for FIR, but it didn't have any effect because FIR metadata serialization does not use the binding context, and should've been simply removed.)